morphir/src/org/finos/morphir/datamodel/Defaults.scala

```diff
@@ -0,0 +1,121 @@
+    case Concept.String            => defaultString
+    case Concept.Order             => defaultOrder
+    case Concept.DayOfWeek         => defaultDayOfWeek
+    case Concept.Char              => defaultChar
+    case Concept.Boolean           => defaultBoolean
+    case Concept.Any               => defaultAny
+    case concept: Concept.Optional => defaultOptional(concept)
+    case concept: Concept.Result   => defaultResult(concept)
+    case concept: Concept.Set      => defaultSet(concept)
+    case concept: Concept.Tuple    => defaultTuple(concept)
+  }
+  def fillWithDefaults(data: Data, concept: Concept): DefaultTask
+  def defaultList(concept: Concept.List): DefaultTask = ZIO.succeed(Data.List.empty(concept.elementType))
+  def defaultMap(concept: Concept.Map): DefaultTask =
+    ZIO.succeed(Data.Map.empty(concept.keyType, concept.valueType))
+  def defaultAlias(concept: Concept.Alias): DefaultTask =
+    default(concept.value).map(Data.Aliased(_, concept))
+  def defaultUnion(concept: Concept.Union) = default(concept.cases.head).map(Data.Union(_, concept))
+  def defaultRecord(concept: Concept.Record): DefaultTask = for {
+    dataFields <- ZIO.collectAll(concept.fields.map { (label, concept) => default(concept).map((label, _)) })
+    res = Data.Record(dataFields, concept)
+  } yield res
+  def defaultStruct(concept: Concept.Struct): DefaultTask = for {
+    dataFields <- ZIO.collectAll(concept.fields.map { (label, concept) => default(concept).map((label, _)) })
+    res = Data.Struct(dataFields)
+  } yield res
+
+  def defaultEnum(concept: Concept.Enum): DefaultTask = {
+    val firstCase = concept.cases.head
+    for {
+      args <- ZIO.collectAll(firstCase.fields.map { (label, concept) => default(concept).map((label, _)) })
+      res = Data.Case(args, firstCase.label.toString(), concept)
+    } yield res
+  }
+  def defaultInt64: DefaultTask   = ZIO.succeed(Data.Int64(0))
+  def defaultInteger: DefaultTask = ZIO.succeed(Data.Integer(BigInt(0)))
+  def defaultDecimal: DefaultTask = ZIO.succeed(Data.Decimal(BigDecimal(0)))
+  def defaultInt16: DefaultTask   = ZIO.succeed(Data.Int16(0))
+  def defaultNothing: DefaultTask = ZIO.fail(NoDefaultNothing())
+  def defaultMonth: DefaultTask   = ZIO.succeed(Data.Month(java.time.Month.JANUARY))
+  def defaultLocalDate: DefaultTask =
+    ZIO.succeed(Data.LocalDate(java.time.LocalDate.EPOCH))
+  def defaultFloat: DefaultTask = ZIO.succeed(Data.Float(0))
+  def defaultLocalTime: DefaultTask =
+    ZIO.succeed(Data.LocalTime(java.time.LocalTime.MIDNIGHT))
+  def defaultInt32: DefaultTask  = ZIO.succeed(Data.Int32(0))
+  def defaultByte: DefaultTask   = ZIO.succeed(Data.Byte(0))
+  def defaultUnit: DefaultTask   = ZIO.succeed(Data.Unit)
+  def defaultString: DefaultTask = ZIO.succeed(Data.String(""))
+  def defaultOrder: DefaultTask  = ZIO.succeed(Data.Order.apply(0))
+  def defaultDayOfWeek: DefaultTask =
+    ZIO.succeed(Data.DayOfWeek(java.time.DayOfWeek.of(0)))
```

Comment:
Why do we need ZIO for this?

Reply:
Was your intention to use ZIO's `Gen` to provide generators?

Reply:
I used ZIO for a few reasons:

- Some of these can already fail (Concept.Nothing should not be able to make a default)
- Even if this implementation always succeeds, users might want to specify failures for some cases - for instance, not creating defaults for some privileged types - so I wanted the interface to allow for such
- Future-proofing (we recently had issues with contracts/exceptions because a function we didn't originally think could fail actually could)
- Consistency with ZIO usage thru-out the project (basically that's just why I used ZIO rather than Either)
-

Reply:
I had no plans regarding ZIO's `Gen`

Reply:
Some thoughts that popped up in my head from the reply:
1) The exception would be the default and would be the same exception either way. It's not a side effect that would require ZIO.
2) Users specifying specific failures defeats the purpose of a default. If users want special logic, they shouldn't be doing that here.
3) Defaults should never side effect.

Reply:
I'm not clear what you mean by 1, can you rephrase?

More generally, I feel like I may have stepped in it by naming this "Default" - the behavior of filling in a `Concept` with some data is something that can call for specific implementations and specific failures. Would you be happier if this were just not named Default?

Reply:
Oh, one specific thing - if these aren't allowed to fail, what should the default for `Concept.Nothing` be?